### PR TITLE
[Feat] Add clip_gradients hook to LightningModule

### LINKED ISSUE #6346: [RFC] Gradient clipping hooks in the LightningModule
## 🚀 Feature

Add clipping hooks to the LightningModule

### Motivation

It's currently very difficult to change the clipping logic

### Pitch

```python
class LightningModule:
    def on_before_clip_gradients():
        ...

    def clip_gradients(self, optimizer):
        ...
    
    def on_after_clip_gradients():
        ...
```

The default implementation would be the same as we currently provide, where the trainer clipping parameters are used.

Maybe those would be deprecated in favor of LightningModule properties.

```python
class LightningOptimizer
    def step(closure=closure)
        if closure is None:
            closure = do_nothing_closure
        def wrapper_closure()
            closure()
            self._trainer.call_hook("on_before_clip_gradients")
            self._trainer.call_hook("clip_gadients")
            self._trainer.call_hook("on_after_clip_gadients")
        self.optimizer.step(closure=wrapper_closure)
```

Need to evaluate the limitations since clipping is currently tied to plugins

### Additional context

This would fix #5096, https://github.com/PyTorchLightning/pytorch-lightning/pull/6123#discussion_r587415267, https://github.com/PyTorchLightning/pytorch-lightning/issues/5671, https://github.com/PyTorchLightning/pytorch-lightning/issues/5982, and allow easily implementing new clipping techniques without having to merge them into Lightning

cc: @rohitgr7 who has been pushing for this for a while

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR adds a new hook clip_gradients user can override to customise the logic to apply for doing gradient clipping.

Currently, this takes 2 arguments, optimizer, optimizer_idx.

Fixes #6346

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Review by ananthsub — COMMENTED** [review]
i don't follow this code flow: we go from loop -> lightning module -> trainer -> accelerator

if we're introducing this hook, shouldn't we go from loop -> accelerator -> lightning module ? where the accelerator calls the lightning module's clip_gradients hook? 

and what do we do about the gradient clip value/algorithm as flags on the trainer? Aren't they ignored if clip gradients is a hook on the module? 

i left some other questions on the linked issue too

**Comment by carmocca** [comment]
Does this close #9584? cc: @rohitgr7

**Comment by rohitgr7** [comment]
@carmocca yes it does.. but I couldn't get time to work on it due to initial setup and onboarding, also was thinking about the points raised by @ananthsub on the issue before moving forward.

So putting some of my initial thoughts here.

We should deprecate these attributes from Trainer and include them inside clip_gradients hook. The reason for keeping them inside clip_gradients is to allow users use the functionalities we already have defined in the backend using super
```py
def clip_gradients(self, optimizer, optimizer_idx, gradient_clip_val, gradient_clip_algorithm):
    if optimizer_idx == 1:
        super().clip_gradients(optimizer, optimizer_idx, gradient_clip_val, gradient_clip_algorithm)
```

while deprecating, make them None inside trainer by default, and don't raise a warning if they are not set and set them to the values they have inside trainer. if user set's them then of-course raise a deprecation warning.

@tchaton should I close mine?

also @ananthsub 
> we go from loop -> lightning module -> trainer -> accelerator

we have a similar flow in `optimizer_step` too.

File: pytorch_lightning/core/lightning.py [hunk 1]
```diff
@@ -1583,6 +1583,30 @@ def optimizer_zero_grad(self, epoch, batch_idx, optimizer, optimizer_idx):
...
+        """Override this method to change the default behaviour of ``clip_gradients``.
+
+        Args:
+            optimizer: Current optimizer
+            optimizer_idx: Current optimizer idx
+            optimizer_idx: If you used multiple optimizers this indexes into that list.
+
+        Examples::
+
+            # DEFAULT
```
Comment (awaelchli) [inline pytorch_lightning/core/lightning.py:1600]: the implementation shown here is not the default

**Comment by tchaton** [comment]
Hey @rohitgr7,

I didn't see you were working on this one already. Let's close mine instead and keep yours.

We need to decide what to do with the Trainer arguments. I still believe they are quite useful for new users.

Best,
T.C

**Comment by four4fish** [comment]
1. gradient_clip_algorithm and gradient_clip_val are params for trainer, should we move them to lightning module too then? 
2. If clip_gradients hook is been overridden, gradient_clip_algorithm and gradient_clip_val set from trainer will not be respected right? Can we make it more clear by adding warning or misconfigError?
